### ISSUE TITLE
TS Error : instanceof Element always return false with Next.js

### ISSUE THREAD

**Issue by purp1eeeee** [remarkablemark/html-react-parser#221]
## Reproducible Demo
https://codesandbox.io/s/quizzical-lake-16q4o?file=/pages/index.tsx

When replacing to a custom component, false will always be returned.

I am currently using the any type.
However, is there any other solution other than the any type?

**Comment by remarkablemark** [comment]
Thanks for opening the issue @Murasaki-1102

In your CodeSandbox example, you're not returning a React element so replace won't work. See [readme](https://github.com/remarkablemark/html-react-parser/tree/v1.2.3#replace).

So can you replace the logs with a return of your custom element?

```diff
export default function IndexPage() {
   const html = "<h1>hoge</h1>";
   const options: HTMLReactParserOptions = {
     replace: (domNode) => {
-      console.log(domNode);
-      console.log(domNode instanceof Element);
+      if (domNode.name === "h1") {
+        return <h1>replaced</h1>;
+      }
     }
   };
   return <div>{parse(html, options)}</div>;
 }
```

See updated [CodeSandbox](https://codesandbox.io/s/html-react-parser-221-2dtui?file=/pages/index.tsx).

**Comment by kakaeriel** [comment]
I following [this instruction](https://github.com/remarkablemark/html-react-parser/tree/v1.2.3#replace), and I also have same problem, I can't get property `attribs` from `domNode`. 

Because typescript check `domNode` is instance of `DOMNode`, not `Element` type, so that the error occurs, because `DOMNode` hasn't `attribs` property. 

And if you check with this condition `console.log(domNode instanceof Element)` the return always false, because instance of domNode is `DOMNode`.

I have solution for this, maybe can help. I defined the `Element` type at the parameter domNode.

```javascript
import parse, { HTMLReactParserOptions } from "html-react-parser";
import { Element } from "domhandler/lib/node";

export function contentHandler(postContent: string) {
  const options: HTMLReactParserOptions = {
    replace: (domNode: Element) => {
      if (domNode.attribs) {
        if (domNode.attribs.id === 'shortcode') {
          return <div className="leadform">Shortcode</div>;
        }
      }
    },
  };

  return parse(postContent, options);
}

```

**Comment by wheredoesyourmindgo** [comment]
I can confirm I'm having the same issue with a Next.js project; Typescript type-casting seems to be the only work around after upgrading html-react-parser.

**Comment by Murasaki-1102** [comment]
@kakaeriel 
so good! thank you👍

**Comment by sarah-j-smith** [comment]
Having these issues with a Typescript project. Like @kakaeriel I'm following this instruction about [replace including the section on Typescript.](https://github.com/remarkablemark/html-react-parser/tree/v1.2.3#replace)

The solution from @kakaeriel works for me. The `instanceof` operator doesn't work (Element is not a type) and after hours browsing through node_modules looking for a type definition, finding this workaround is a life-saver.

**Comment by remarkablemark** [comment]
Thanks for providing an alternative solution @kakaeriel. Would you be interested in opening a PR to improvement the `README.md` documentation?

**Comment by kakaeriel** [comment]
@remarkablemark Yes sure! I will create a PR, ASAP.

**Comment by Naxos84** [comment]
Wow. I have the same issue using NextJS.


> 
> Because typescript check `domNode` is instance of `DOMNode`, not `Element` type, so that the error occurs, because `DOMNode` hasn't `attribs` property.
> 
> ```

So why does this only occur in a NextJS app.
I setup a react-scripts typescript app and there it works.
https://codesandbox.io/s/html-parsing-and-replacing-elements-yjtv7?file=/src/index.tsx

**Comment by remarkablemark** [comment]
I'm not sure @Naxos84. I bet Next.js has a different TypeScript configuration. Please check out @kakaeriel's [solution](https://github.com/remarkablemark/html-react-parser/issues/221#issuecomment-771600574) for a workaround.

@kakaeriel let me know if you're able to open a PR to update the `README.md` with your workaround. If you're unable to do it, I can do it. Thanks!

**Comment by Naxos84** [comment]
I don't think that it is NextJs related.
I tried a little more and voila it worked somehow.
I changed @remarkablemark codesandbox and added `console.log(domElement.constructor)` as mentioned on [Stack Overflow](https://stackoverflow.com/questions/13613524/get-an-objects-class-name-at-runtime).

The result looks as follows:
![grafik](https://user-images.githubusercontent.com/17813025/108814476-25e08900-75b3-11eb-9afa-b0560c006234.png)

So I don't understand why it is suddenly working!?

**Comment by Naxos84** [comment]
I suggest using a custom typeguard.

```
const isElement = (domNode: DOMNode): domNode is Element => {
    const isTag = domNode.type === "tag";
    const hasAttributes = (domNode as Element).attribs !== undefined;

    return isTag && hasAttributes;
};
```

```
const parserOptions: HTMLReactParserOptions = {
    replace: domNode => {
        if (isElement(domNode)) {
            return <p>It's an element.</p>;
        } else {
            return <p>It's something else</p>;
        }
    },
};
```

**Comment by natterstefan** [comment]
Hey everyone, I was able to fix this in Next.js@10.x with [this solution](https://gist.github.com/natterstefan/3bc712eca6ff88781d687b7240a78cc1). It's a public Gist where you can find my current implementation details.

**Comment by remarkablemark** [comment]
@natterstefan Thanks for sharing your solution!

**Comment by vladislav-puzyrev** [comment]
I have the same problem with NextJS. Element is always undefined. I typed domNode as any and converted the options type as HTMLReactParserOptions. Element from domhandler doesn't work with SSR?

```typescript
const options = {
  replace: (domNode: any) => {
      // ...
  }
} as HTMLReactParserOptions;
```

**Comment by roybcr** [comment]
I was able to fix this with the code below, hope it would help

<img width="891" alt="Screen Shot 2021-06-12 at 20 32 31" src="https://user-images.githubusercontent.com/68123426/121784745-6265de00-cbbe-11eb-8999-feb5f505aea2.png">

**Comment by remarkablemark** [comment]
#296 released in [1.4.0](https://github.com/remarkablemark/html-react-parser/pull/306)

**Comment by tomconroy** [comment]
It's possible to use duck typing to get domNode as an Element:

```ts
    parse(html, {
      replace: (domNode) => {
        if ('name' in domNode && 'attribs' in domNode) {
          // domNode is an Element
        }
      },
    });
```
